File: solutions/python/all-your-base/1/all_your_base.py
def rebase(input_base, digits, output_base):
    if input_base < 2:
        raise ValueError("input base must be >= 2")
    if output_base < 2:
        raise ValueError("output base must be >= 2")
    if any(d < 0 or d >= input_base for d in digits):
        raise ValueError("all digits must satisfy 0 <= d < input base")
    value = 0
    for d in digits:
        value = value * input_base + d
    if value == 0:
        return [0]
    result = []
    while value:
        result.append(value % output_base)
        value //= output_base
    return result[::-1]

File: solutions/javascript/all-your-base/1/all-your-base.js
const digitsToNumber = (digits, base) => {
  let num = 0;
  for (const d of digits) {
    num = num * base + d;
  }
  return num;
};

const numberToDigits = (num, base) => {
  if (num === 0) return [0];
  const digits = [];
  while (num > 0) {
    digits.unshift(num % base);
    num = Math.floor(num / base);
  }
  return digits;
};

export const convert = (digits, inputBase, outputBase) => {
  if (inputBase < 2) throw new Error('Wrong input base');
  if (outputBase < 2) throw new Error('Wrong output base');
  if (digits.length === 0) throw new Error('Input has wrong format');
  if (digits.length > 1 && digits[0] === 0) throw new Error('Input has wrong format');
  for (const d of digits) {
    if (d < 0 || d >= inputBase) throw new Error('Input has wrong format');
  }
  const num = digitsToNumber(digits, inputBase);
  return numberToDigits(num, outputBase);
};
